lib/containers/accumulators/review-comments-accumulator.js

```diff
@@ -0,0 +1,122 @@
+import React from 'react';
+import PropTypes from 'prop-types';
+import {graphql, createPaginationContainer} from 'react-relay';
+
+import {PAGE_SIZE, PAGINATION_WAIT_TIME_MS} from '../../helpers';
+import {RelayConnectionPropType} from '../../prop-types';
+import Accumulator from './accumulator';
+
+export class BareReviewCommentsAccumulator extends React.Component {
+  static propTypes = {
+    // Relay props
+    relay: PropTypes.shape({
+      hasMore: PropTypes.func.isRequired,
+      loadMore: PropTypes.func.isRequired,
+      isLoading: PropTypes.func.isRequired,
+    }).isRequired,
+    reviewThread: PropTypes.shape({
+      comments: RelayConnectionPropType(
+        PropTypes.object,
+      ),
+    }),
+
+    // Render prop. Called with (error or null, array of all review comments, loading)
+    children: PropTypes.func,
+
+    // Called right after refetch happens
+    onDidRefetch: PropTypes.func.isRequired,
+  }
+
+  render() {
+    const resultBatch = this.props.reviewThread.comments.edges.map(edge => edge.node);
+
+    return (
+      <Accumulator
+        relay={this.props.relay}
+        resultBatch={resultBatch}
+        onDidRefetch={this.props.onDidRefetch}
+        pageSize={PAGE_SIZE}
+        waitTimeMs={PAGINATION_WAIT_TIME_MS}>
+        {(error, comments, loading) => this.props.children({error, comments, loading})}
+      </Accumulator>
+    );
+  }
+}
+
+export default createPaginationContainer(BareReviewCommentsAccumulator, {
+  reviewThread: graphql`
+    fragment reviewCommentsAccumulator_reviewThread on PullRequestReviewThread
+    @argumentDefinitions(
+      commentCount: {type: "Int!"}
+      commentCursor: {type: "String"},
+    ) {
+      id
+      comments(
+        first: $commentCount
+        after: $commentCursor
+      ) @connection(key: "ReviewCommentsAccumulator_comments") {
+        pageInfo {
+          hasNextPage
+          endCursor
+        }
+
+        edges {
+          cursor
+          node {
```

Comment:
This also applies to the other two accumulators: `ReviewSummariesAccumulator` and `ReviewThreadsAccumulator`

Caveat: In my head this seems like it would be a straight-forward improvement, but there might be complexities I'm not quite seeing yet.

Reply:
And to be clear, I don't consider this ship-blocking.

Reply:
I started to try to address this and I hit a bit of a hitch with it. The aggregated data structure that we assemble in `AggregatedReviewsContainer` to hold the collated review threads and review comments is different from the way that the data is shaped in GraphQL, so we can't easily map it to props in a Relay fragmentContainer around `ReviewsView`.

Specifically, the GraphQL structure looks roughly like this:

```graphql
... on PullRequest {
  reviewThreads {
    nodes {
      # fields we need from review threads
      ...reviewsView_reviewThreads
      comments {
        nodes {
          # fields we need from comments
          ...reviewsView_comments
        }
      }
    }
  }
}
```

We process comments grouped with their owning threads in `ReviewsView` like so:

```js
const commentThreads = [
  {thread: {id: 'thread0'}, comments: [{id: 'comment0'}, {id: 'comment1'}]},
  {thread: {id: 'thread1'}, comments: [{id: 'comment2'}, {id: 'comment3'}]},
]

<ReviewsView commentThreads={commentThreads} />
```

But as our components are structured now, Relay would provide the fragment data to `ReviewsView` like this:

```js
const reviewThreads = [
  {id: 'thread0'}, {id: 'thread1'},
]

const comments = [
  {id: 'comment0'}, {id: 'comment1'}, {id: 'comment2'}, {id: 'comment3'},
]

<ReviewsView reviewThreads={reviewThreads} comments={comments} />
```

So we'd lose the ability to tell which comment belongs to which review thread, which we use pretty heavily within the view.

I think one path forward to accomplish this is to split a subcomponent out of ReviewsView which can be a fragmentContainer for a single comment:

```graphql
... on PullRequest {
  reviewThreads {
    nodes {
      # fields we need from review threads
      ...reviewsView_reviewThread
      comments {
        nodes {
          # fields we need from comments
          ...reviewCommentView_comment
        }
      }
    }
  }
}
```

```js
class ReviewsView extends React.Component {
  renderReviewThread(thread) {
    // Relay will do its magic and allow each child to access each comment's data, but ReviewsView still gets to tell
    // it about its owning thread
    return thread.comments.map(comment => (
      <ReviewCommentView
        comment={comment}
        thread={thread}
      />
    ))
  }
}
```

In any case, this was a bit more invasive than I was hoping it would be, so I'm going to punt on it for the moment, and tackle this after we have a chance to get some of the easier ones out of the way, and maybe advocate merging as-is if we continue to run short on time. It's also possible that I'm misunderstanding some Relay-fu that would make this easier. I welcome education if so :grin:

Reply:
Filed as #2046 :rocket: